README.md
# konsole: readable, pleasing console output

> When you are writing a Python command line tool and your head is on fire
> because of overly rich frameworks that just don’t click.

![An interactive Python session using
konsole](https://raw.githubusercontent.com/apparebit/konsole/boss/session.png)

[konsole](https://github.com/apparebit/konsole) is a simple logger built on top
of Python's `logging` framework that prints to standard error and, if the
underlying terminal is amenable to it, does so with the judicious use of bold
and light type as well as a dash of color. This package's interface stands on
its own, no experience or direct interaction with `logging` required. At the
same time, this package plays equally well with other loggers, just leave
~~konsole~~ 🙄 console output to it.


## Using konsole

In order to use konsole, you need to go through the usual motions of installing

```shell
(venv) project % python3 -m pip install konsole
```

and then importing the package

```python
import konsole
```

konsole automatically integrates itself with Python’s logging system the first
time the module is imported into an application. Notably, it registers a handler
that prints messages to standard error with the root logger, replaces the
current logger class with a subclass that supports the `detail` keyword
argument, and enables the capture of Python warnings through the logging system.

konsole's public API follows below. It consists of one function to update the
configuration, one function to access the `__main__` application logger, and six
functions to print messages at different priority levels. konsole includes type
annotations, which have been validated with
[mypy](https://mypy.readthedocs.io/en/stable/).


### Configuring konsole

  * Change the minimum level for printing messages and/or the flag for forcing
    colors on/off.

    ```python
    def config(
        *,
        level: Optional[int] = None,
        use_color: Optional[bool] = None,
        volume: Optional[int] = None,
    ) -> None: ...
    ```

    konsole starts out with `INFO` as minimum level and uses color if
    standard error is a TTY. The `volume` argument provides an alternative
    means for setting the output level, with larger volumes printing more
    information and 0 corresponding to the `W


### Logging Messages

  * Get the `__main__` application logger. konsole uses it for writing messages.

    ```python
    def logger() -> logging.Logger
    ```

    The logger, like any other logger created after the initialization of
    konsole, supports the `detail` keyword argument (see below).

  * Log a message at the given level.

    ```python
    def critical(msg: str, *args: object, **kwargs: object) -> None: ...
    def error(msg: str, *args: object, **kwargs: object) -> None: ...
    def warning(msg: str, *args: object, **kwargs: object) -> None: ...
    def info(msg: str, *args: object, **kwargs: object) -> None: ...
    def debug(msg: str, *args: object, **kwargs: object) -> None: ...
    def log(level: int, msg: str, *args: object, **kwargs: object) -> None: ...
    ```

    The message string is the first and only mandatory argument. If the message
    string contains `%` format specifiers, the necessary values must follow as
    positional arguments.

    Valid keyword arguments include those supported by Python's logging
    framework, notably `exc_info` for including an exception's stacktrace. They
    also include `detail` for supplemental data. konsole prints the mapping,
    sequence, or scalar value on separate, indented lines after the message but
    before an exception's stacktrace.

    konsole defines ALL CAPS constants, e.g., `WARNING`, for the five levels
    above. They have the same values as the corresponding constants in Python's
    logging package.


---

© 2022 [Robert Grimm](https://apparebit.com).
Subject to [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0) license.
On [GitHub](https://github.com/apparebit/konsole).
On [PyPI](https://pypi.org/project/konsole/).


CHANGELOG.md
# konsole changelog

### v0.7.0 (15 Jun 2022):

  * Add support for `volume` argument to `config()` as convenient alternative to
    `level` argument for command line tools.

### v0.6.0 (29 Mar 2022):

  * After converting arbitrary detail values to strings, also split them into
    lines.
  * Ensure that type annotations are treated as strings, so that generic types
    do not trip up konsole in Python 3.7 or 3.8.

### v0.5.1 (7 Feb 2022):

  * Fix `__all__` to list only symbols that actually exist.

### v0.5.0 (7 Feb 2022):

  * Use custom logging handler to make output redirectable with `contextlib`'s
    `redirect_stderr`.
  * Remove the now superfluous `konsole.redirect()`.

### v0.4.0 (3 Feb 2022):

  * Fix unexpected exception upon detail being an empty dictionary.
  * More generally, improve handling of detail values that are effectively empty.
  * Fix indentation of first line of exception trace.
  * Add screenshot illustrating konsole's output to [project readme](README.md)

### v0.3.0 (29 Jan 2022):

  * Initialize konsole eagerly on first import. Replace `init()` with `config()`.
  * Instead of replacing logging configuration with `basicConfig()`, augment
    current configuration by adding handler to root logger and by replacing
    logger class with subclass.
  * Remove runtime dependencies on the standard library's `collections.abc`,
    `dataclasses`, and `textwrap` packages to reduce import latency and memory
    overhead.
  * Completely refactor test script, improve coverage.
  * Update documentation accordingly.

### v0.2.0 (26 Jan 2022):

  * Surface `detail` keyword argument for logging functions.
  * Document konsole's public API.
  * Correctly handle corner cases of `detail` being `None`,
    `set_color()` and `set_level()` calling `init()`.
  * Cache main logger and directly access flag to eliminate repeated, trivial
    function calls.

### v0.1.0: 25 Jan 2022

Initial release.


test.py
from __future__ import annotations
from contextlib import redirect_stderr
from io import StringIO
import logging
import re

# konsole does not use a test runner such as pytest for two reasons: First,
# bringing in a major dependency for relatively little testing code seems
# overkill. Second, pytest in particular captures all console output itself and
# thus interferes with konsole's redirection of standard error.


def test_konsole() -> None:
    # Validate pre-konsole-import configuration.
    vanilla_logger = logging.getLogger("vanilla_logger")
    assert not logging.getLogger().hasHandlers()
    assert logging.getLogger().level == logging.WARNING

    import konsole

    # Continue validating pre-konsole-import configuration.
    assert isinstance(vanilla_logger, logging.Logger)
    assert not isinstance(vanilla_logger, konsole.KonsoleLogger)

    # Validate post-konsole-import configuration.
    assert logging.getLogger().hasHandlers()
    assert logging.getLogger().level == konsole.INFO

    local_logger = logging.getLogger(__file__)
    assert isinstance(konsole.logger(), konsole.KonsoleLogger)
    assert isinstance(local_logger, konsole.KonsoleLogger)

    buffer = StringIO()
    with redirect_stderr(buffer):
        konsole.config(use_color=True)
        konsole.info("fyi")
        konsole.error("bad!", detail="broken!")

        konsole.config(use_color=False)
        konsole.critical("big %s!", "bad")
        konsole.debug("yo!")  # no output.

        konsole.config(level=konsole.DEBUG)
        local_logger.warning("beware!", detail=665)
        konsole.info("look", detail="one\ntwo")
        konsole.debug("detail", detail={})

    # Validate konsole output.
    actual, expected = prepare(buffer.getvalue())
    compare(actual, expected)
    assert actual == expected

    # Levels and volumes. NB: Current level is DEBUG.
    konsole.config(level=konsole.ERROR)
    assert logging.getLogger().level == konsole.ERROR
    konsole.config(level=konsole.ERROR, volume=2)
    assert logging.getLogger().level == konsole.DEBUG
    konsole.config(volume=-665)
    assert logging.getLogger().level == konsole.CRITICAL
    konsole.config(volume=665)
    assert logging.getLogger().level == konsole.DEBUG


ESC_SEQ = re.compile(r"\x1b\[(?P<code>\d+(?:;\d+)*)m")


def prepare(output: str) -> tuple[list[str], list[str]]:
    actual = re.sub(ESC_SEQ, "⊲\g<code>⊳", output).split("\n")

    # fmt: off
    s = Style("⊲", "⊳")
    expected = "".join([
        s.info("[INFO]"), s.space, s.message("fyi."), s.newline,
        s.error("[ERROR]"), s.space, s.message("bad!"),
        s.detail("\n    broken!"), s.newline,
        '[CRITICAL] big bad!', s.newline,
        '[WARNING] beware!', s.newline,
        '    665', s.newline,
        '[INFO] look:', s.newline,
        '    one', s.newline,
        '    two', s.newline,
        '[DEBUG] detail.', s.newline,
    ]).split("\n")
    # fmt: on

    return actual, expected


def compare(actual: list[str], expected: list[str]) -> None:
    # SGR is NOT part of konsole's public API!
    import konsole

    red = konsole.SGR("31", "39")
    green = konsole.SGR("32", "39")

    for actual_line, expected_line in zip(actual, expected):
        if actual_line == expected_line:
            print("  " + actual_line)
        else:
            print(red("- " + actual_line))
            print(green("+ " + expected_line))


# Helper class to keep expected output readable by independently applying styles.
class Style:
    newline = "\n"
    space = " "

    def __init__(self, pre: str = "\x1b[", post: str = "m") -> None:
        self.pre = pre
        self.post = post

    def critical(self, _s_: str) -> str:
        return f"{self.pre}1;35{self.post}{_s_}{self.pre}0;39{self.post}"

    def detail(self, _s_: str) -> str:
        return f"{self.pre}90{self.post}{_s_}{self.pre}0{self.post}"

    def error(self, _s_: str) -> str:
        return f"{self.pre}1;31{self.post}{_s_}{self.pre}0;39{self.post}"

    def exception(self, _s_: str) -> str:
        return f"{self.pre}90{self.post}{_s_}{self.pre}0{self.post}"

    def info(self, _s_: str) -> str:
        return f"{self.pre}1{self.post}{_s_}{self.pre}0{self.post}"

    def message(self, _s_: str) -> str:
        return f"{self.pre}1{self.post}{_s_}{self.pre}0{self.post}"

    def warning(self, _s_: str) -> str:
        return f"{self.pre}1;38;5;208{self.post}{_s_}{self.pre}0;39{self.post}"


if __name__ == "__main__":
    test_konsole()


konsole/__init__.py
"""
konsole: readable, pleasing console output

[konsole](https://github.com/apparebit/konsole) is a simple logger built on top
of Python's `logging` framework that prints to standard error and, if the
underlying terminal is amenable to it, does so with the judicious use of bold
and light type as well as a dash of color. This package's interface stands on
its own, no experience or direct interaction with `logging` required. At the
same time, this package plays equally well with other loggers, just leave
console output to it.
"""
from __future__ import annotations

import logging
import sys

from typing import Any, cast, Optional, TextIO, TYPE_CHECKING

__version__ = "0.7.0"

__all__ = [
    'config', 'logger',
    'CRITICAL', 'ERROR', 'WARNING', 'INFO', 'DEBUG',
    'critical', 'error', 'warning', 'info', 'debug', 'log',
]


class _NoSuchValueType:
    pass


_NoSuchValue = _NoSuchValueType()
del _NoSuchValueType


CRITICAL = logging.CRITICAL
ERROR = logging.ERROR
WARNING = logging.WARNING
INFO = logging.INFO
DEBUG = logging.DEBUG

VOLUME_TO_LEVEL = {
    -2: CRITICAL,
    -1: ERROR,
    0: WARNING,
    1: INFO,
    2: DEBUG,
}


# --------------------------------------------------------------------------------------


class SGR:
    """
    The Select Graphic Rendition (SGR) for enabling and disabling a style but
    without the prefix of escape character and opening square bracket '[' as
    well as the suffix of small letter 'm'. SGR are not expected to nest.
    """

    def __init__(self, on: str, off: str) -> None:
        self._on = on
        self._off = off

    def __call__(self, text: str) -> str:
        return f"\x1b[{self._on}m{text}\x1b[{self._off}m"


class KonsoleFormatter(logging.Formatter):
    """
    A log formatter targeting the console for human consumption. If supported by
    the terminal, this class uses indentation, bold and light type, as well as a
    dash of color to make log output look less busy and easier to read.
    """

    DETAIL = "<detail>"
    EXCEPTION = "<exception>"
    MESSAGE = "<message>"

    PUNCTUATION = ("!", ",", ".", ":", ";", "?", '"', "'")

    STYLE = {
        # Styles for level labels:
        "CRITICAL": SGR("1;35", "0;39"),
        "ERROR": SGR("1;31", "0;39"),
        "WARNING": SGR("1;38;5;208", "0;39"),
        "INFO": SGR("1", "0"),
        # Styles for other components of log entries:
        MESSAGE: SGR("1", "0"),
        DETAIL: SGR("90", "0"),
        EXCEPTION: SGR("90", "0"),
    }

    def __init__(self, use_color: Optional[bool] = None) -> None:
        super().__init__("%(message)s")
        self.use_color = use_color if use_color is not None else sys.stderr.isatty()

    def applyStyle(self, name: str, text: str) -> str:
        style = self.STYLE.get(name)
        return style(text) if style and self.use_color else text

    def formatDetail(self, record: logging.LogRecord) -> str:
        detail = getattr(record, "detail", _NoSuchValue)
        if detail is _NoSuchValue:
            return ""

        lines: list[str]
        if isinstance(detail, dict):
            width = max((len(key) for key in detail), default=0)
            lines = [f"\n    {key:>{width}} = {value}" for key, value in detail.items()]
        elif isinstance(detail, (list, tuple, set, frozenset)):
            lines = [f"\n    {item}" for item in detail]
        else:
            lines = [f"\n    {line}" for line in str(detail).split("\n")]

        text = "".join(lines)
        return self.applyStyle(self.DETAIL, text) if text else ""

    def formatMessage(self, record: logging.LogRecord) -> str:
        levelname = self.applyStyle(record.levelname, f"[{record.levelname}]")
        message = super().formatMessage(record)
        detail = self.formatDetail(record)

        if not message.endswith(self.PUNCTUATION):
            # Ensure that message consistently ends with appropriate punctuation.
            message += ":" if detail or record.exc_info else "."
        if levelname[0] == "\x1b":
            # Only highlight message if level is highlighted too. Debug is just plain.
            message = self.applyStyle(self.MESSAGE, message)
        if detail and record.exc_info:
            # Both detail and exception are formatted the same, need separating line.
            detail += "\n\n"

        return f"{levelname} {message}{detail}"

    def formatException(self, exc_info: Any) -> str:
        text = super().formatException(exc_info)
        text = "\n    ".join(text.split("\n"))
        return self.applyStyle(self.EXCEPTION, f"    {text}")


class KonsoleHandler(logging.StreamHandler):
    # Always dereference sys.stderr so that konsole's output is redirectable.

    @property  # type: ignore
    def stream(self) -> TextIO:  # type: ignore
        return sys.stderr

    @stream.setter
    def stream(self, stream: TextIO) -> None:
        pass


# --------------------------------------------------------------------------------------


# Instead of subclassing logging.Logger directly, we create the konsole logger
# as a subclass of the currently configured logger class. That way, konsole
# composes with other extensions to the logging framework. Since mypy doesn't
# handle this intermingling of types and values, we also define a more static
# view — while type checking.

if TYPE_CHECKING:
    Logger = logging.Logger
else:
    Logger = logging.getLoggerClass()


class KonsoleLogger(Logger):
    # Unfortunately, the various logging methods of logging.Logger directly call
    # into Logger._log() if isEnabledFor(LEVEL) and _log() does *not* accept
    # arbitrary keyword arguments. Hence adding support for the detail keyword
    # is a bit more involved than it should be.

    def critical(self, msg: object, *args: object, **kwargs: object) -> None:
        self.log(CRITICAL, msg, *args, **kwargs)

    def error(self, msg: object, *args: object, **kwargs: object) -> None:
        self.log(ERROR, msg, *args, **kwargs)

    def warning(self, msg: object, *args: object, **kwargs: object) -> None:
        self.log(WARNING, msg, *args, **kwargs)

    def info(self, msg: object, *args: object, **kwargs: object) -> None:
        self.log(INFO, msg, *args, **kwargs)

    def debug(self, msg: object, *args: object, **kwargs: object) -> None:
        self.log(DEBUG, msg, *args, **kwargs)

    def log(self, level: int, msg: object, *args: object, **kwargs: Any) -> None:
        if "detail" in kwargs:
            _prepare_detail(kwargs)
        super().log(level, msg, *args, **kwargs)


def _prepare_detail(kwargs: dict[str, object]) -> None:
    detail = kwargs.pop("detail")
    if "extra" not in kwargs:
        kwargs.update(extra={"detail": detail})
        return

    extra = kwargs["extra"]
    if not isinstance(extra, dict):
        raise ValueError(f'"extra" must be a dictionary but is "{extra}"!')

    extra.update(detail=detail)


# --------------------------------------------------------------------------------------


# Initialize konsole when this module is first imported by selectively changing
# the configuration of Python's logging module.

_formatter = KonsoleFormatter()
_handler = KonsoleHandler()
_handler.setFormatter(_formatter)
logging.setLoggerClass(KonsoleLogger)
_main_logger = cast(KonsoleLogger, logging.getLogger("__main__"))

logging.captureWarnings(True)
logging.getLogger().addHandler(_handler)
logging.getLogger().setLevel(INFO)


def config(
    *,
    level: Optional[int] = None,
    use_color: Optional[bool] = None,
    volume: Optional[int] = None,
) -> None:
    """
    Optionally set the logging level and forcibly enable or disable color.

    The volume argument provides a convenient alternative way of specifying the
    level — with larger volumes increasing the amount of information logged and
    zero corresponding to `WARNING`. If specified, the volume takes priority.
    """
    if isinstance(volume, int):
        level = VOLUME_TO_LEVEL[max(-2, min(volume, 2))]
    if level is not None:
        logging.getLogger().setLevel(level)
    if use_color is not None:
        _formatter.use_color = use_color


# --------------------------------------------------------------------------------------


def logger() -> logging.Logger:
    """Get the `__main__` application logger."""
    return _main_logger


def critical(msg: object, *args: object, **kwargs: object) -> None:
    """Log the given critical message."""
    _main_logger.log(CRITICAL, msg, *args, **kwargs)


def error(msg: object, *args: object, **kwargs: Any) -> None:
    """Log the given error message."""
    _main_logger.log(ERROR, msg, *args, **kwargs)


def warning(msg: object, *args: object, **kwargs: Any) -> None:
    """Log the given warning message."""
    _main_logger.log(WARNING, msg, *args, **kwargs)


def info(msg: object, *args: object, **kwargs: Any) -> None:
    """Log the given message."""
    _main_logger.log(INFO, msg, *args, **kwargs)


def debug(msg: object, *args: object, **kwargs: Any) -> None:
    """Log the given debug message."""
    _main_logger.log(DEBUG, msg, *args, **kwargs)


def log(level: int, msg: object, *args: object, **kwargs: Any) -> None:
    """Log the given message at the given level."""
    _main_logger.log(level, msg, *args, **kwargs)
